coderd/coderd.go

```diff
@@ -1238,13 +1238,25 @@ func (api *API) CreateInMemoryProvisionerDaemon(ctx context.Context, name string
 				if xerrors.Is(err, io.EOF) {
 					return
 				}
-				logger.Debug(ctx, "drpc server error", slog.Error(err))
+				logger.Debug(dialCtx, "drpc server error", slog.Error(err))
 			},
 		},
 	)
+	// in-mem pipes aren't technically "websockets" but they have the same properties as far as the
+	// API is concerned: they are long-lived connections that we need to close before completing
+	// shutdown of the API.
+	api.WebsocketWaitMutex.Lock()
+	api.WebsocketWaitGroup.Add(1)
+	api.WebsocketWaitMutex.Unlock()
 	go func() {
-		err := server.Serve(ctx, serverSession)
-		logger.Info(ctx, "provisioner daemon disconnected", slog.Error(err))
+		defer api.WebsocketWaitGroup.Done()
+		// here we pass the background context, since we want the server to keep serving until the
+		// client hangs up.  If we, say, pass the API context, then when it is canceled, we could
+		// drop a job that we locked in the database but never passed to the provisionerd.  The
+		// provisionerd is local, in-mem, so there isn't a danger of losing contact with it and
+		// having a dead connection we don't know the status of.
+		err := server.Serve(context.Background(), serverSession)
```

Comment:
Sanity check: will this result in provisioner jobs not being cancelled on shutdown? And is that really what we want?

Let's say a provisioner job (terraform) is running and it will take 5 minutes, I imagine we would like to ensure that job is cancelled gracefully, so we only wait for the cleanup but not the full 5 minutes. Is that what will happen or not?

Reply:
in cli/server.go we explicitly close the provisionerDaemons before closing the API.  Closing the provisioner daemon fails the active job.  That's not exactly graceful, but at least it leaves the job in an understandable state.